[SQL] LEFT ASOF JOIN on string columns produces error[E0308]: mismatched types

Following SQL:

```
CREATE TABLE tbl1(
id INT, c2 VARCHAR);

CREATE TABLE tbl2(
id INT, c2 VARCHAR);

CREATE MATERIALIZED VIEW v AS SELECT
i.id, i.c2 AS i_c2, v.c2 AS v_c2
FROM tbl1 i
LEFT ASOF JOIN tbl2 v
MATCH_CONDITION ( i.c2 >= v.c2 )
ON i.id = v.id;
```
Given example input **{data1}** for **{tbl1}** and **{data2}** for **{tbl2}**:
```
data1 = [{"id": 1, "c2": "apple"}]
data2 = [{"id": 1, "c2": "bye"}]
```

Produces the following error:
```
 error[E0308]: mismatched types
  --> crates/feldera_pipe_operator_5ed9155dda38a911/src/lib.rs:90:13
   |
89 |         Option<SqlString> {
   |         ----------------- expected `std::option::Option<feldera_sqllib::SqlString>` because of return type
90 |             ((*p0).get()).1.as_ref()
   |             ^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<SqlString>`, found `Option<&SqlString>`
   |
   = note: expected enum `std::option::Option<feldera_sqllib::SqlString>`
              found enum `std::option::Option<&feldera_sqllib::SqlString>`
help: try removing the method call
   |
90 -             ((*p0).get()).1.as_ref()
90 +             ((*p0).get()).1
   |
```
```
 error[E0308]: mismatched types
  --> crates/feldera_pipe_operator_5ed9155dda38a911/src/lib.rs:94:13
   |
93 |         Option<SqlString> {
   |         ----------------- expected `std::option::Option<feldera_sqllib::SqlString>` because of return type
94 |             ((*p0).get()).0.as_ref()
   |             ^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<SqlString>`, found `Option<&SqlString>`
   |
   = note: expected enum `std::option::Option<feldera_sqllib::SqlString>`
              found enum `std::option::Option<&feldera_sqllib::SqlString>`
help: try removing the method call
   |
94 -             ((*p0).get()).0.as_ref()
94 +             ((*p0).get()).0
   |
```

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/LowerAsof.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/LowerAsof.java
@@ -218,8 +218,12 @@ public class LowerAsof implements CircuitTransform {
             DBSPType commonTSType = leftTSType.withMayBeNull(nullable);
             Utilities.enforce(commonTSType.sameType(rightTSType.withMayBeNull(nullable)));
 
-            DBSPClosureExpression leftTimestamp = leftTS.cast(leftTS.getNode(), commonTSType, false).closure(leftVar);
-            DBSPClosureExpression rightTimestamp = rightTS.cast(rightTS.getNode(), commonTSType, false).closure(rightVar);
+            DBSPClosureExpression leftTimestamp = leftTS.cast(leftTS.getNode(), commonTSType, false)
+                    .applyCloneIfNeeded()
+                    .closure(leftVar);
+            DBSPClosureExpression rightTimestamp = rightTS.cast(rightTS.getNode(), commonTSType, false)
+                    .applyCloneIfNeeded()
+                    .closure(rightVar);
 
             DBSPVariablePath k = keyType.ref().var();
             DBSPVariablePath l0 = wrappedLeftType.ref().var();

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
@@ -1199,5 +1199,20 @@ public class Regression1Tests extends SqlIoTest {
                 roww NOT IN (roww) AS roww
                 FROM tbl;""");
     }
+
+    @Test
+    public void issue4975() {
+        this.getCCS("""
+                CREATE TABLE tbl1(id INT, c2 VARCHAR);
+                
+                CREATE TABLE tbl2(id INT, c2 VARCHAR);
+                
+                CREATE MATERIALIZED VIEW v AS SELECT
+                i.id, i.c2 AS i_c2, v.c2 AS v_c2
+                FROM tbl1 i
+                LEFT ASOF JOIN tbl2 v
+                MATCH_CONDITION ( i.c2 >= v.c2 )
+                ON i.id = v.id;""");
+    }
 }
  

